### PR TITLE
Raise a deprecation warning when a field is annotated as final with a default value

### PR DESCRIPTION
**Pull request by Viicos**
Also adapt the `collect_model_fields` logic a bit.

Part of https://github.com/pydantic/pydantic/issues/11119.

## Checklist

* [ ] The pull request title is a good summary of the changes - it will be used in the changelog
* [ ] Unit tests for the changes exist
* [ ] Tests pass on CI
* [ ] Documentation reflects the changes where applicable
* [ ] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

File: pydantic/warnings.py [hunk 1]
```diff
@@ -6,8 +6,11 @@
 
 __all__ = (
     'PydanticDeprecatedSince20',
-    'PydanticDeprecationWarning',
```
Comment (sydney-runkle) [inline pydantic/warnings.py:9]: We should probably still export this, might be used in user code, and we export in `__init__.py`?

File: tests/test_main.py [hunk 2]
```diff
@@ -2011,8 +2012,10 @@ class Model(BaseModel):
     ids=['no-arg', 'with-arg'],
 )
 def test_final_field_decl_with_default_val(ann):
```
Comment (sydney-runkle) [inline tests/test_main.py:2014]: Maybe note this is deprecated in the test name?

File: pydantic/_internal/_fields.py [hunk 3]
```diff
@@ -158,7 +159,18 @@ def collect_model_fields(  # noqa: C901
...
+        assigned_value = getattr(cls, ann_name, PydanticUndefined)
+
+        if _is_finalvar_with_default_val(ann_type, assigned_value):
+            warnings.warn(
+                f'Annotation {ann_name!r} is marked as final and has a default value. Pydantic treats {ann_name!r} as a '
+                f'class variable, but it will be considered as an normal field in V3. If you still want {ann_name!r} to be '
+                'considered as a class variable, annotate it as: `ClassVar[<type>] = <default>.`',
+                category=PydanticDeprecatedSince211,
+                # Incorrect when `create_model` is used, but the chance that final with a default is used is low in that case:
+                stacklevel=4,
```
Comment (sydney-runkle) [inline pydantic/_internal/_fields.py:172]: I'm fine with this for now, should be easy enough to change if we get an issue report.

File: pydantic/_internal/_fields.py [hunk 4]
```diff
@@ -158,7 +159,18 @@ def collect_model_fields(  # noqa: C901
...
             continue
-        if _is_finalvar_with_default_val(ann_type, getattr(cls, ann_name, PydanticUndefined)):
+
+        assigned_value = getattr(cls, ann_name, PydanticUndefined)
+
+        if _is_finalvar_with_default_val(ann_type, assigned_value):
+            warnings.warn(
+                f'Annotation {ann_name!r} is marked as final and has a default value. Pydantic treats {ann_name!r} as a '
+                f'class variable, but it will be considered as an normal field in V3. If you still want {ann_name!r} to be '
+                'considered as a class variable, annotate it as: `ClassVar[<type>] = <default>.`',
```
Comment (sydney-runkle) [inline pydantic/_internal/_fields.py:169]: Can we link to the typing spec change that motivated this change?

**Review by sydney-runkle — CHANGES_REQUESTED** [review]
Nice work on the refactor using `assigned_value` instead of re-fetching depending on the branch you hit.

LGTM overall, I left a few minor change requests.

**Comment by sydney-runkle** [comment]
Do we know why we treated finalvars with defaults as classvars in the past?

File: pydantic/warnings.py (hunk 1 above)
Comment (Viicos) [reply pydantic/warnings.py:9]: It is still exported, just reordered.

File: pydantic/_internal/_fields.py (hunk 4 above)
Comment (Viicos) [reply pydantic/_internal/_fields.py:169]: I added "to be aligned with dataclasses" as the typing spec change was actually just an update to follow what stdlib dataclasses where doing.

**Comment by Viicos** [comment]
> Do we know why we treated finalvars with defaults as classvars in the past?

As per [the PR](https://github.com/pydantic/pydantic/pull/2768#issue-882561425) that implemented this:

> Implementation is based on a `PEP 591`:
> > Type checkers should infer a final attribute that is initialized in a class body as being a class variable.

**Review by sydney-runkle — APPROVED** [review]
Feedback looks good

**Pull request merged**